### PR TITLE
[Model] Add BNB quantization support for Mllama

### LINKED ISSUE #9714: [Bug]: AttributeError: Model MllamaForConditionalGeneration does not support BitsAndBytes quantization yet
### Your current environment

Collecting environment information...
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.31

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-122-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA H100 80GB HBM3

### Model Input Dumps

`vllm serve unsloth/Llama-3.2-90B-Vision-Instruct-bnb-4bit --quantization bitsandbytes --load-format bitsandbytes --trust-remote-code --enforce-eager`

```
Initializing an LLM engine (v0.6.3.post1) with config: model='unsloth/Llama-3.2-90B-Vision-Instruct-bnb-4bit', speculative_config=None
, tokenizer='unsloth/Llama-3.2-90B-Vision-Instruct-bnb-4bit', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, rope_scaling=None,
rope_theta=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.bfloat16, max_seq_len=131072, download_dir=None, load_format=LoadFormat.BITSANDBYTES, tensor_pa
rallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=bitsandbytes, enforce_eager=True, kv_cache_dtype=auto, quantization_param_path=None, d
evice_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forw
ard_time=False, collect_model_execute_time=False), seed=0, served_model_name=unsloth/Llama-3.2-90B-Vision-Instruct-bnb-4bit, num_scheduler_steps=1, chunked_prefill_enabled=F
alse multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=False, use_cached_outputs=True, mm_processor_kwargs=None)
```

### 🐛 Describe the bug

`AttributeError: Model MllamaForConditionalGeneration does not support BitsAndBytes quantization yet`

I was trying the Llama-3.2-90B-Vision-Instruct-bnb-4bit model, it shows such an error. Not sure which place is better to raise this issue, [unsloth](https://github.com/unslothai/unsloth) or [transformers](https://github.com/huggingface/transformers) or just here.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Isotr0py**
FILL IN THE PR DESCRIPTION HERE

FIX #9714 (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/mllama.py [hunk 1]
```diff
@@ -795,19 +803,17 @@ def attention_with_mask(
...
-                                              kv_len,
-                                              self.head_dim).contiguous()
+                                              kv_len, self.head_dim)
         v = v.transpose(0,
                         1)[:,
                            None, :, :].expand(self.num_local_key_value_heads,
                                               self.num_key_value_groups,
-                                              kv_len,
-                                              self.head_dim).contiguous()
+                                              kv_len, self.head_dim)
```
Comment (mgoin) [inline vllm/model_executor/models/mllama.py:816]: Why did you remove this? I added this intentionally to avoid illegal memory access in SDPA with batched multi-image

Comment (Isotr0py) [reply vllm/model_executor/models/mllama.py]: Ooops, I didn't realize this when I merged from my another old fork to support BNB. :(

Have fixed this.

**Review by mgoin — APPROVED** [review]
LGTM, have you tested loading a model on HF?

**Comment by Isotr0py** [comment]
Sure. I ran both [unsloth/Llama-3.2-11B-Vision-Instruct-bnb-4bit](https://huggingface.co/unsloth/Llama-3.2-11B-Vision-Instruct-bnb-4bit) and [Isotr0py/Llama-3.2-11B-Vision-Instruct-bnb-8bit](https://huggingface.co/Isotr0py/Llama-3.2-11B-Vision-Instruct-bnb-8bit) with full/partial quantization. Both model can be loaded and generate reasonable outputs:
<details>
<summary>Llama-3.2-11B-Vision-Instruct-bnb-4bit</summary>

```text
INFO 10-29 07:02:53 llm_engine.py:240] Initializing an LLM engine (v0.1.dev3049+g82c2515.d20241019) with config: model='unsloth/Llama-3.2-11B-Vision-Instruct-bnb-4bit', speculative_config=None, tokenizer='unsloth/Llama-3.2-11B-Vision-Instruct-bnb-4bit', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, rope_scaling=None, rope_theta=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.float16, max_seq_len=4096, download_dir=None, load_format=LoadFormat.BITSANDBYTES, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=bitsandbytes, enforce_eager=True, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=unsloth/Llama-3.2-11B-Vision-Instruct-bnb-4bit, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=False, use_cached_outputs=False, chat_template_text_format=string, mm_processor_kwargs=None)
INFO 10-29 07:02:54 enc_dec_model_runner.py:141] EncoderDecoderModelRunner requires XFormers backend; overriding backend auto-selection and forcing XFormers.
INFO 10-29 07:02:54 selector.py:119] Using XFormers backend.
INFO 10-29 07:03:04 model_runner.py:1055] Starting to load model unsloth/Llama-3.2-11B-Vision-Instruct-bnb-4bit...
INFO 10-29 07:03:05 selector.py:119] Using XFormers backend.
INFO 10-29 07:03:05 loader.py:1064] Loading weights with BitsAndBytes quantization.  May take a while ...
INFO 10-29 07:03:05 weight_utils.py:243] Using model weights format ['*.safetensors']
model-00002-of-00002.safetensors: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.18G/2.18G [00:52<00:00, 41.9MB/s]
model-00001-of-00002.safetensors: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5.00G/5.00G [01:59<00:00, 41.9MB/s]
Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]███████████████████████████████████████████████████████████████████████████████| 5.00G/5.00G [01:59<00:00, 42.1MB/s]
Loading safetensors checkpoint shards:  50% Completed | 1/2 [00:01<00:01,  1.82s/it]
Loading safetensors checkpoint shards: 100% Completed | 2/2 [00:02<00:00,  1.16s/it]
Loading safetensors checkpoint shards: 100% Completed | 2/2 [00:02<00:00,  1.26s/it]

Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  50% Completed | 1/2 [00:02<00:02,  2.37s/it]
Loading safetensors checkpoint shards: 100% Completed | 2/2 [00:03<00:00,  1.47s/it]
Loading safetensors checkpoint shards: 100% Completed | 2/2 [00:03<00:00,  1.60s/it]

INFO 10-29 07:05:11 model_runner.py:1066] Loading model weights took 6.7244 GB
INFO 10-29 07:05:11 enc_dec_model_runner.py:301] Starting profile run for multi-modal models.
INFO 10-29 07:05:52 worker.py:260] Memory profiling results: total_gpu_memory=14.74GiB initial_memory_usage=6.88GiB peak_torch_memory=7.74GiB memory_usage_post_profile=6.93Gib non_torch_memory=0.20GiB kv_cache_size=5.33GiB gpu_memory_utilization=0.90
INFO 10-29 07:05:52 gpu_executor.py:122] # GPU blocks: 2183, # CPU blocks: 1638
INFO 10-29 07:05:52 gpu_executor.py:126] Maximum concurrency for 4096 tokens per request: 8.53x
WARNING 10-29 07:05:59 preprocess.py:89] Falling back on <BOS> for decoder start token id because decoder start token id is not available.
Processed prompts: 100%|██████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:58<00:00, 14.75s/it, est. speed input: 0.68 toks/s, output: 4.34 toks/s]
 The image depicts a tall tower with a white dome and a spire, surrounded by pink cherry blossom trees. The tower is in the center of the image, and the blossoms are in the foreground. The sky is blue and clear, suggesting that it is spring or early summer. The overall atmosphere of the image is
 The image is a photograph of a white tower framed by cherry blossoms. The tower is in the center of the image and is framed by branches of cherry blossoms. The tower is tall and white with a pointed top. The cherry blossoms are pink and are in full bloom. The background is a clear blue sky
 The image depicts a tall tower, likely a skyscraper or a tower of some kind, with a white exterior and a pointed top, surrounded by a tree with pink flowers. The tree is in the foreground of the image, and the tower is in the background. The sky is blue and clear, suggesting that the photo
 The image shows a tall white tower, framed by branches of a cherry blossom tree. The tower is in the center of the image, and it is surrounded by pink flowers. The sky is blue and clear. The overall atmosphere suggests a spring or summer day, with the cherry blossoms in full bloom. The image is
```

</details>


<details>
<summary>Llama-3.2-11B-Vision-Instruct-bnb-8bit</summary>

```text
WARNING 10-29 13:47:25 config.py:1707] Casting torch.bfloat16 to torch.float16.
WARNING 10-29 13:47:29 config.py:364] bitsandbytes quantization is not fully optimized yet. The speed can be slower than non-quantized models.
WARNING 10-29 13:47:29 config.py:438] To see benefits of async output processing, enable CUDA graph. Since, enforce-eager is enabled, async output processor cannot be used
INFO 10-29 13:47:29 llm_engine.py:240] Initializing an LLM engine (vdev) with config: model='/root/autodl-tmp/Llama-3.2-11B-Vision-Instruct-bnb-8bit', speculative_config=None, tokenizer='/root/autodl-tmp/Llama-3.2-11B-Vision-Instruct-bnb-8bit', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, rope_scaling=None, rope_theta=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.float16, max_seq_len=4096, download_dir=None, load_format=LoadFormat.BITSANDBYTES, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=bitsandbytes, enforce_eager=True, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=/root/autodl-tmp/Llama-3.2-11B-Vision-Instruct-bnb-8bit, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=False, use_cached_outputs=False, chat_template_text_format=string, mm_processor_kwargs=None)
INFO 10-29 13:47:30 enc_dec_model_runner.py:141] EncoderDecoderModelRunner requires XFormers backend; overriding backend auto-selection and forcing XFormers.
INFO 10-29 13:47:30 selector.py:119] Using XFormers backend.
/root/miniconda3/lib/python3.12/site-packages/xformers/ops/fmha/flash.py:211: FutureWarning: `torch.library.impl_abstract` was renamed to `torch.library.register_fake`. Please use that instead; we will remove `torch.library.impl_abstract` in a future version of PyTorch.
  @torch.library.impl_abstract("xformers_flash::flash_fwd")
/root/miniconda3/lib/python3.12/site-packages/xformers/ops/fmha/flash.py:344: FutureWarning: `torch.library.impl_abstract` was renamed to `torch.library.register_fake`. Please use that instead; we will remove `torch.library.impl_abstract` in a future version of PyTorch.
  @torch.library.impl_abstract("xformers_flash::flash_bwd")
INFO 10-29 13:47:31 model_runner.py:1055] Starting to load model /root/autodl-tmp/Llama-3.2-11B-Vision-Instruct-bnb-8bit...
INFO 10-29 13:47:31 selector.py:119] Using XFormers backend.
INFO 10-29 13:47:31 loader.py:1064] Loading weights with BitsAndBytes quantization.  May take a while ...
Loading safetensors checkpoint shards:   0% Completed | 0/3 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  33% Completed | 1/3 [00:00<00:01,  1.70it/s]
Loading safetensors checkpoint shards:  67% Completed | 2/3 [00:02<00:01,  1.19s/it]
Loading safetensors checkpoint shards: 100% Completed | 3/3 [00:03<00:00,  1.38s/it]
Loading safetensors checkpoint shards: 100% Completed | 3/3 [00:03<00:00,  1.27s/it]

Loading safetensors checkpoint shards:   0% Completed | 0/3 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  33% Completed | 1/3 [00:00<00:01,  1.72it/s]
Loading safetensors checkpoint shards:  67% Completed | 2/3 [00:02<00:01,  1.22s/it]
Loading safetensors checkpoint shards: 100% Completed | 3/3 [00:03<00:00,  1.40s/it]
Loading safetensors checkpoint shards: 100% Completed | 3/3 [00:03<00:00,  1.29s/it]

INFO 10-29 13:47:39 model_runner.py:1066] Loading model weights took 11.0580 GB
INFO 10-29 13:47:39 enc_dec_model_runner.py:301] Starting profile run for multi-modal models.
/root/miniconda3/lib/python3.12/site-packages/bitsandbytes/autograd/_functions.py:316: UserWarning: MatMul8bitLt: inputs will be cast from torch.bfloat16 to float16 during quantization
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")
INFO 10-29 13:47:51 worker.py:260] Memory profiling results: total_gpu_memory=31.73GiB initial_memory_usage=11.50GiB peak_torch_memory=13.64GiB memory_usage_post_profile=13.43Gib non_torch_memory=1.07GiB kv_cache_size=13.85GiB gpu_memory_utilization=0.90
INFO 10-29 13:47:51 gpu_executor.py:122] # GPU blocks: 5671, # CPU blocks: 1638
INFO 10-29 13:47:51 gpu_executor.py:126] Maximum concurrency for 4096 tokens per request: 22.15x
WARNING 10-29 13:47:54 preprocess.py:89] Falling back on <BOS> for decoder start token id because decoder start token id is not available.
Processed prompts: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [01:18<00:00, 19.75s/it, est. speed input: 0.51 toks/s, output: 3.24 toks/s]
 The image shows a cherry blossom tree in front of a tall tower. The tree is in full bloom, with pink flowers covering its branches. The tower is white and has a distinctive shape, with a large dome at the top and a series of smaller domes below it. The sky is blue and clear, suggesting that
 The image is a photograph of a white tower framed by cherry blossoms. The tower is in the center of the image and is tall and slender. It has a round top and a narrow, tall, white, metal lattice structure. The tower is framed by branches of a blooming cherry tree. The branches are in
 The image is a photograph of a white tower surrounded by pink cherry blossoms. The tower is in the center of the image and is tall and slender, with a rounded top. It has a series of windows and a spire at the top. The tower is surrounded by branches of a cherry blossom tree that are covered
 The image depicts a tall white tower, possibly a skyscraper or a tower, surrounded by pink cherry blossoms. The tower is in the center of the image and is surrounded by branches of pink cherry blossoms. The sky is blue and clear, suggesting that it is a sunny day. The overall atmosphere of the image
```
</details>

**Comment by luizhsalazar** [comment]
Is this feature available in which version? or is still in dev?
